simple CLI interface for all commands using M_CLI2

urbanjost opened the pull request:
An implimentation of a CLI interface based on the most
recent master for discussion with @everythingfunctional

certik commented:
Thanks!

On Sat, Sep 19, 2020, at 5:01 PM, urbanjost wrote:
> An implimentation of a CLI interface based on the most
> recent master for discussion with @everythingfunctional 
> <https://github.com/everythingfunctional>
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/fortran-lang/fpm/pull/181
> 
> Commit Summary
> 
>  * simple CLI interface for all commands using M_CLI2
> File Changes
> 
>  * *M* fpm/fpm.toml 
> <https://github.com/fortran-lang/fpm/pull/181/files#diff-7b7f15ad033448e9fa834a26171257de> (3) 
>  * *M* fpm/src/fpm.f90 
> <https://github.com/fortran-lang/fpm/pull/181/files#diff-e235e41ae1a801d1ff11bfa3aa1ea6db> (133) 
>  * *M* fpm/src/fpm_command_line.f90 
> <https://github.com/fortran-lang/fpm/pull/181/files#diff-7fc26592b0babf0895d006b372f1ec37> (192) 
> Patch Links:
> 
>  * https://github.com/fortran-lang/fpm/pull/181.patch
>  * https://github.com/fortran-lang/fpm/pull/181.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/fortran-lang/fpm/pull/181>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAAFAWDWKG2RRQMA37KSDADSGUS37ANCNFSM4RTICRKQ>.
>

urbanjost commented:
It says it fails on a file that was a working file on windows that ends in ".f90.".  I added the dot to keep it from being used, which worked on Linux.  I deleted the file, it is just fodder for a new demo. Interesting it picked up that file even though it did not end in ".f90". Not sure if how the checks work; if it will pull a new dependency and run again or not.

LKedward commented:
Thanks @urbanjost. The error was a Windows/git error; it is illegal for a filename to end in a dot on Windows so the git checkout failed. I've rerun the checks and all seems fine now.

I will look over this PR today

LKedward commented on fpm/fpm.toml:
```diff
@@ -10,6 +10,9 @@ copyright = "2020 fpm contributors"
 tag = "v0.2"
 
+[dependencies.M_CLI2]
+git = "https://github.com/urbanjost/M_CLI2.git"
```
```suggestion
git = "https://github.com/urbanjost/M_CLI2.git"
commit = "a177b0077819571815fa6a8da6980bcb45443858"
```

We need to pin our dependencies to a specific commit or release. Here I've chosen the most recent from M_CLI2.

urbanjost replied:
The keyword commit= works with the Haskell fpm but when I try that with the toml-f is says commit= is not allowed. Otherwise I agree with this.  Should that be a separate issue?  Or should I add a tag? I have no problem with just including the M_CLI2.f90 file but I was following the (really nice) toml-f implementation

LKedward replied:
Apologies, this is my mistake. It should be `rev` instead of `commit`.
```suggestion
git = "https://github.com/urbanjost/M_CLI2.git"
rev = "a177b0077819571815fa6a8da6980bcb45443858"
```

urbanjost replied:
It is actually more typical to do all the command line parsing in one place with M_CLI2, but the arguments can be called from anywhere. Since all the model settings and such were being set and used in fpm I just parsed and provided help text in the command_line procedure and retrieved them in the cmd routines where they were used. It is trivial to move the calls to get_args(3f) back into command_line; just not sure whether it means adding options to the settings or to the calls. I am looking at the code trying to see what the model is there.

LKedward commented on fpm/src/fpm.f90:
```diff
@@ -25,23 +29,36 @@ subroutine build_model(model, settings, package, error)
+             & -fPIC &
+             & -fmax-errors=1 &
+             & -ffast-math &
+             & -funroll-loops '
```
```suggestion
             & -funroll-loops ' // &
             & '-J'//join_path(model%output_directory,model%package_name)
```

LKedward commented on fpm/src/fpm.f90:
```diff
@@ -114,21 +131,111 @@ subroutine cmd_build(settings)
+     stop 4
+   endif
+
+   message=[character(len=80) :: &                                           ! create fpm.toml
```
```suggestion
   message=[&                                           ! create fpm.toml
```

I don't think you need to specify `character(len=...)` here - it's actually a little misleading with arbitrary length `80`. And similarly below.

LKedward reviewed (commented):
Thanks for this @urbanjost. As you've mentioned elsewhere it would be good to get the CLI implemented to move things along in other areas.
I've left a few comments but my main feedback is that I think all CLI processing should be contained to `fpm_command_line.f90`.

My interpretation of the current setup is that all argument processing would occur in `fpm_command_line.f90` with the results being stored in the various `cmd_settings` types. These types are then passed as arguments to routines in `fpm.f90` so that no further interrogation of the command line is required. Therefore M_CLI2 would only be `use`'d in `fpm_command_line.f90`.

Is this possible with M_CLI2 or is this approach not a good idea?
